### DISCUSSION TITLE
How do you reuse actions amongst controllers, that inherit from another base class?

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/aspnetcore#55253] [Q&A]
Let's say we have a generic base class:

```
public class ListController<T> : Controller where T class, new()
```

Then we inherit two controllers from this `ListController`

```
public class CustomerController : ListController<Customer>
```
and
```
public class OrderController : ListController<Order>
```

Now let's say I have a common action, that can not be put inside the `ListController`. How do you share that action between these controllers without repeating yourself?

**Comment by markmouton22** [comment]
Hi :)

One way to share a common action between these controllers without repeating yourself is to use an **interface** or an **abstract** **base** **class** that defines this common action.


```
public interface ICommonAction
{
    ActionResult CommonAction();
}
```

```
public class CustomerController : ListController<Customer>, ICommonAction
{
    public ActionResult CommonAction()
    {
        // Implement the common action here
    }
}
```

```
public class OrderController : ListController<Order>, ICommonAction
{
    public ActionResult CommonAction()
    {
        // Implement the common action here
    }
}
```
In this way, both CustomerController and OrderController have the CommonAction method, and you can call it on any instance of these classes that you know implements ICommonAction.

**Reply by Nefcanto** [reply]
@markmouton22, that's not reusing. That's a contract. Maybe you can call the default methods a reuse. But that's also extremely limited, as it does not have access to HttpContext or it can not be used with class-level variables (instance variables). There is very little value in implementing contracts as a method of DRY and code reuse.
